### PR TITLE
MCO-1898: MCS serves image-aware first-boot config

### PR DESCRIPTION
**Pull request by dkhater-redhat**
**- What I did**

1. Taught the MCS to write image annotations at bootstrap.
2. Extended appendNodeAnnotations(...) in pkg/server/server.go to optionally include:
   a. `machineconfiguration.openshift.io/currentImage`
   b. `machineconfiguration.openshift.io/desiredImage`
   c. alongside the existing MC/state keys when generating the initial node-annotations file (/etc/machine-config-daemon/node-annotations.json).
3. Added the image into the node-annotations appender.
   a. The server can now pass the resolved rendered OS image (e.g., MOSC/MOSB output) into appendNodeAnnotations(...) so new nodes start with authoritative image annotations.

Impact:
New nodes pivot/validate directly against the intended layered image during bootstrap. In image mode, new nodes successfully deploy the rendered image and can complete without an extra reboot when no reboot-requiring changes are present.

**- How to verify it**

1. Create a Quay push secret in the MCO namespace
2. Apply a `MachineOSConfig` with `renderedImagePushSpec` pointing at your Quay repo/tag.
3. Opt nodes into image mode (e.g., via the pool opt-in label/annotation you use).
5. Scale up a worker so a new node is provisioned:
`oc scale machineset.machine.openshift.io -n openshift-machine-api dkhater-10-15-2025-a-cpb2n-worker-us-east-1c --replicas=1`
6. Observe the new node- it should pull and deploy the rendered image without reboot 
7. Confirm on the node:

```
oc debug node/ip-10-0-85-7.ec2.internal
chroot /host
cat /etc/machine-config-daemon/currentimage
rpm-ostree status
```

Expected:

1. `/etc/machine-config-daemon/currentimage` contains your Quay digest.
2. `rpm-ostree status` shows the same digest as the booted deployment.
3. No node reboot occurs

Example:

```
$ cat /etc/machine-config-daemon/currentimage
quay.io/dkhater/image-mode-testing@sha256:d9dcf52c08a659561c36bd62a0a57fcec34aec2f71bce88cd68fd314ddf8db70
```

```
$ rpm-ostree status
* ostree-unverified-registry:quay.io/dkhater/image-mode-testing@sha256:d9dcf52c08a659561c36bd62a0a57fcec34aec2f71bce88cd68fd314ddf8db70
  Digest: sha256:d9dcf52c08a659561c36bd62a0a57fcec34aec2f71bce88cd68fd314ddf8db70
  Version: 9.6.20251013-1
```
_please note, if you use the internal image registry, you will see the legacy two node boots occur_
**- Description for the changelog**

MCS now embeds current/desired image annotations in the initial node annotations at bootstrap. This makes the MCD pivot/validate directly against the rendered layered OS image. New nodes can complete image-mode provisioning without an unnecessary reboot when no reboot-requiring changes are present.

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -1721,11 +1722,13 @@ func (dn *Daemon) getStateAndConfigs() (*stateAndConfigs, error) {
 		klog.Infof("%s is not set. any errors? %s", constants.CurrentImageAnnotationKey, err)
 		return nil, err
 	}
+	klog.Infof("[daemon.go] getStateAndConfigs: currentImage=%s (from node annotations)", currentImage)
 	desiredImage, err := getNodeAnnotationExt(dn.node, constants.DesiredImageAnnotationKey, true)
 	if err != nil {
 		klog.Infof("%s is not set. any errors? %s", constants.DesiredImageAnnotationKey, err)
 		return nil, err
 	}
+	klog.Infof("[daemon.go] getStateAndConfigs: desiredImage=%s (from node annotations)", desiredImage)
```
Comment (djoshy) [inline pkg/daemon/daemon.go]: same as previous

File: pkg/server/cluster_server.go [hunk 2]
```diff
@@ -163,7 +174,27 @@ func (cs *clusterServer) GetConfig(cr poolRequest) (*runtime.RawExtension, error
...
+		appendInitialMachineConfig,
+		func(cfg *ign3types.Config, _ *mcfgv1.MachineConfig) error { return appendCerts(cfg, []string{}, "") },
+		// Inject desired image into the MC
+		appendDesiredOSImage(desiredImage),
+		// Must be last
+		func(cfg *ign3types.Config, mc *mcfgv1.MachineConfig) error {
+			return appendEncapsulated(cfg, mc, cr.version)
+		},
+	}
+
```
Comment (djoshy) [inline pkg/server/cluster_server.go:206]: Ahh, wish we could pretty this up :sweat:  but I guess we had to refactor these since we need to pass in new arguments now? More of a nit :smile:

File: pkg/daemon/daemon.go [hunk 3]
```diff
@@ -2174,8 +2177,19 @@ func (dn *Daemon) checkStateOnFirstRun() error {
...
+		// During bootstrap, prefer the image from node annotations (if set) over the MC from cluster lister
+		targetOSImageURL := state.currentImage
+		if targetOSImageURL == "" {
+			// Fall back to MC's OSImageURL if no image annotation was provided
+			targetOSImageURL = state.currentConfig.Spec.OSImageURL
+		}
+		klog.Infof("[daemon.go] checkStateOnFirstRun: in bootstrap mode, bootedOSImageURL=%s, targetOSImageURL=%s (from state.currentImage annotation or state.currentConfig.Spec.OSImageURL)", dn.bootedOSImageURL,
+			targetOSImageURL)
 		osMatch := dn.checkOS(targetOSImageURL)
+		klog.Infof("[daemon.go] checkStateOnFirstRun: osMatch=%v (bootedOSImageURL == targetOSImageURL)", osMatch)
```
Comment (djoshy) [inline pkg/daemon/daemon.go]: same as previous

File: pkg/server/cluster_server.go [hunk 4]
```diff
@@ -224,3 +255,95 @@ func kubeconfigFromSecret(secretDir, apiserverURL string, caData []byte) ([]byte
...
+		if config.Spec.MachineConfigPool.Name == pool.Name {
+			mosc = config
+			break
+		}
+	}
+
+	// No MOSC for this pool - not layered
+	if mosc == nil {
+		return ""
+	}
```
Comment (djoshy) [inline pkg/server/cluster_server.go:302]: I wonder if we could simplify this, since the MOSC name should always match the pool Name per https://github.com/openshift/api/pull/2399?

File: pkg/server/cluster_server.go [hunk 5]
```diff
@@ -224,3 +255,95 @@ func kubeconfigFromSecret(secretDir, apiserverURL string, caData []byte) ([]byte
...
+			build.Spec.MachineConfig.Name == currentConf {
+			mosb = build
+			break
+		}
+	}
+
+	if mosb == nil {
+		klog.Infof("Pool %s has MachineOSConfig but no matching MachineOSBuild for MC %s", pool.Name, currentConf)
+		return ""
+	}
```
Comment (djoshy) [inline pkg/server/cluster_server.go:350]: Couldn't we use the [MOSB ref field](https://github.com/openshift/api/blob/master/machineconfiguration/v1/types_machineosconfig.go#L122) from the MOSC object directly in the lister? Why do we need to step through the whole list? Sorry if I'm missing something here!

File: pkg/server/cluster_server.go [hunk 6]
```diff
@@ -329,6 +330,16 @@ func (cs *clusterServer) resolveDesiredImageForPool(pool *mcfgv1.MachineConfigPo
 	}
 
 	imageSpec := moscState.GetOSImage()
+
+	// Don't serve internal registry images during bootstrap because cluster DNS is not available yet
+	// New nodes will bootstrap with base image, then update to layered image after joining cluster (2 reboots)
+	// External registries (Quay, etc.) will still get the 1-reboot optimization
+	if strings.Contains(imageSpec, ".svc:") || strings.Contains(imageSpec, ".svc.cluster.local") {
```
Comment (djoshy) [inline pkg/server/cluster_server.go]: I wonder if there is a cluster resource we can pull this value from. https://github.com/openshift/api/blob/master/imageregistry/v1/types.go#L60 exists, but I'm not sure what values they're typically populated with :thinking:

**Review by djoshy — COMMENTED** [review]
Generally looks good! Just left a few questions/cleanups that we could do :smile: 

Could we also squash the second commit?

File: pkg/server/cluster_server.go (hunk 4 above)
Comment (isabella-janssen) [reply pkg/server/cluster_server.go:302]: > since the MOSC name should always match the pool Name

Could their be MOSCs "grandfathered in" that were created prior to that validation change that have a name not matching the MCP name?

File: pkg/server/cluster_server.go (hunk 2 above)
Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:206]: yes i have the same wishes. i tried reusing it but it broke bootstrap because of the new args

File: pkg/server/cluster_server.go (hunk 4 above)
Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:302]: yep you're so right

Comment (djoshy) [reply pkg/server/cluster_server.go:302]: > > since the MOSC name should always match the pool Name
>
> Could their be MOSCs "grandfathered in" that were created prior to that validation change that have a name not matching the MCP name?

Good point! I think it was backported, so I suppose someone could thread the needle between GA and there :sweat_smile: 

Looking at https://issues.redhat.com//browse/OCPBUGS-60904, looks we have 4.18.21-4.18.23 streams where someone could've done that technically

Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:302]: I can keep as is and clean up down the line. Ill add a todo. this code is going to need more work anyways due to the dns resolution issues.

Comment (djoshy) [reply pkg/server/cluster_server.go:302]: Works for me!

Comment (isabella-janssen) [reply pkg/server/cluster_server.go:302]: > Ill add a todo.

When you add that, can you include the Jira story link in either this thread or the code comment, please?

File: pkg/server/cluster_server.go (hunk 5 above)
Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:350]: You're right that we could simplify this, but there's an important nuance here. The current-machine-os-build annotation does track the "latest successful build" but the code in cluster_server.go needs to find the MOSB for a specific rendered MachineConfig that depends on the pool's rollout state. The key issue is that there can be multiple MOSBs for a single MOSC (one per rendered MachineConfig), and the code in cluster_server.go needs to find the MOSB for a specific rendered config (either pool.Spec.Configuration.Name or pool.Status.Configuration.Name).

For example, during a rollout you might have:

worker-abc123 (MOSB for old rendered-worker-1) - successful
worker-def456 (MOSB for new rendered-worker-2) - newly successful
During a rollout, there could be two successful MOSBs (one for the old config, one for the new), and mosc.Status.MachineOSBuild would point to the newer one. But if no nodes have started updating, we need to serve the image for the old config, not the new one.

So if someone scaled a node during a MOSB update (pre-node boot), this would be a safer approach.

File: pkg/server/cluster_server.go (hunk 6 above)
Comment (dkhater-redhat) [reply pkg/server/cluster_server.go]: yeah i personally dont love my approach but it looks like the image registry config resource primarily contains backend configuration stuff? The current approach of checking for .svc: or .svc.cluster.local is a heuristic to detect internal registry URLs. although... looking at this made me realize we have this in the reconciler:

```
❯ rg "getInternalRegistryHostnames" 
pkg/controller/build/reconciler.go
1042:// getInternalRegistryHostnames discovers OpenShift internal registry hostnames
```

let me look into this...

Comment (dkhater-redhat) [reply pkg/server/cluster_server.go]: yeah nvmd we dont have access to the kubeclient at that point. this is what makes this such a fun problem XD

**Comment by dkhater-redhat** [comment]
> Generally looks good! Just left a few questions/cleanups that we could do 😄
> 
> Could we also squash the second commit?

yes! my plan is to squash everything when the team is ready to give an LGTM (incase i need to go back and debug) 😄

File: pkg/server/cluster_server.go (hunk 6 above)
Comment (djoshy) [reply pkg/server/cluster_server.go]: Not via that helper no, but we could potentially set up RBAC so the server has access to this resource? (similar to how you've done for OCL resources)

Comment (dkhater-redhat) [reply pkg/server/cluster_server.go]: see https://redhat-internal.slack.com/archives/GH7G2MANS/p1760717268595939 for more info on this- I will look into using this method

File: pkg/server/cluster_server.go (hunk 5 above)
Comment (djoshy) [reply pkg/server/cluster_server.go:350]: ack, that makes sense to me! I guess in this case, the node would go through another reboot, once the new image build is complete. I suppose it doesn't make sense to wait to serve the new node until the MOSB is done building?

Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:350]: im not sure that i can block it at the boostrap level

File: pkg/controller/common/registry_utils.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+// GetInternalRegistryHostnames discovers OpenShift internal registry hostnames
+// by querying Services and Routes in the openshift-image-registry namespace.
+func GetInternalRegistryHostnames(ctx context.Context, kubeclient clientset.Interface, routeclient routeclientset.Interface) ([]string, error) {
+	var hostnames []string
+
+	// Get the list of services in the openshift-image-registry namespace (cluster-local)
+	services, err := kubeclient.CoreV1().Services("openshift-image-registry").List(ctx, metav1.ListOptions{})
+	if err != nil {
```
Comment (pablintino) [inline pkg/controller/common/registry_utils.go:20]: May this fail if the registry is not deployed? I'd say so if the NS doesn't exist. If so, we should not fail and just asume we don't need to worry about the Image Registry corner case.

File: pkg/controller/common/registry_utils.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
+package common
```
Comment (pablintino) [inline pkg/controller/common/registry_utils.go:1]: Can we UT test the logic introduced here with fakeClients?

File: pkg/server/cluster_server.go (hunk 4 above)
Comment (djoshy) [reply pkg/server/cluster_server.go:302]: > > Ill add a todo.
>
> When you add that, can you include the Jira story link in either this thread or the code comment, please?

Could we also implement Isabella's ask here, so we don't lose track of it?

File: pkg/server/cluster_server.go (hunk 5 above)
Comment (djoshy) [reply pkg/server/cluster_server.go:350]: I don't think I follow - couldn't the MCS just not respond to the new node until the image is ready? I'd imagine the new node will just keep sending requests till it gets a response from the MCS? 

I'm also okay with handling this case as a follow-up, its just not clear to me how serving an older image is any better than serving this node a legacy MachineConfig. It would just result in an additional reboot for the node once the build is complete.

File: pkg/controller/common/registry_utils.go (hunk 7 above)
Comment (djoshy) [reply pkg/controller/common/registry_utils.go:20]: Hmm, in the current form(cb5982e81704dd5812e6aefc19bd28b10b48457a), any errors in the image registry check should return false for  `IsOpenShiftRegistry`, which should result in the `OSImageURL` injection to take place?  I would say this case is already being handled, unless I'm missing something here!

File: pkg/controller/common/registry_utils.go (hunk 8 above)
Comment (dkhater-redhat) [reply pkg/controller/common/registry_utils.go:1]: yes thats the plan!

File: pkg/controller/common/registry_utils.go (hunk 7 above)
Comment (dkhater-redhat) [reply pkg/controller/common/registry_utils.go:20]: we are at a safe default- when the base image is served, we go back to 2 node reboot default. thank you for double checking 💯

File: pkg/server/cluster_server.go (hunk 5 above)
Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:350]: from my experience with this, it would block the node scale-up. i havent monitored the node logs closely in this case (during scale up) but when i did (during debugging) I never saw any retries. the console would just stop showing logs. so intuitively speaking, if we have the MCS block waiting for the new MOSB finish, we will time out and the node boot wil fail. This is not something im confident about and will need to follow up with jerry/try it out myself.

File: pkg/server/cluster_server.go (hunk 4 above)
Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:302]: just saw this. will do!

File: pkg/server/cluster_server.go (hunk 5 above)
Comment (djoshy) [reply pkg/server/cluster_server.go:350]: Agree, I think its worth experimenting in a different card, its definitely not a blocking issue for me!

In my experience, the node should continually try if the MCS doesn't respond to its requests. If the MCS does serve it something that is not usable(such as internal registry image), it will bork the node, yeah. And I also don't think we'd want to the MCS block while processing the request until the build is complete; it can silently exit the request(perhaps we just log for debugging) if the currently ref'd MOSB is building. When the build does complete later, a node request after that can be responded to. I might be making some bad assumptions here, so it is definitely worth testing :sweat_smile:

Comment (yuqi-zhang) [reply pkg/server/cluster_server.go:350]: I'd like to pitch something slightly different, which is that we should potentially match the existing behaviour of node scale-up for non-on cluster image mode nodes.

For the "regular" MCS, we serve the new version iff the new render successfully rolled out to one node. To match this behaviour, we should potentially increase the requirement the other way and only roll out the build if it's successful and any node has updated to it.

This will cause more reboots generally but comes with safety for node scaling while updates are happening. We should do it as a followup regardless, but just wanted to add that there.

Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:350]: thank you @yuqi-zhang - @djoshy, if its ok with you, im going to write up and test the proposed thoughts around this and, if successful, add to a follow-up PR. This work (as-is) will get the core functionality in, and we can add the modifications later (considering this is for an edge case[s]). LMK if that sounds ok with you?

File: pkg/controller/common/registry_utils.go (hunk 8 above)
Comment (dkhater-redhat) [reply pkg/controller/common/registry_utils.go:1]: @pablintino done!

File: pkg/server/cluster_server.go (hunk 5 above)
Comment (djoshy) [reply pkg/server/cluster_server.go:350]: ack, I think that makes sense to me! That does make it cleanly fit into our existing methodology. 

@dkhater-redhat Could I trouble you to write a briefer version [of this comment](https://github.com/openshift/machine-config-operator/pull/5357#discussion_r2440788000) above this block of code to explain why we don't directly use the MOSB ref? I know it would be helpful for future me :smile: And perhaps make a card for the follow-up work and link it there as a TODO too?

EDIT: @dkhater-redhat ninja'd me, but I'm good with your proposal!

File: pkg/controller/build/reconciler.go [hunk 9]
```diff
@@ -999,7 +999,7 @@ func (b *buildReconciler) deleteMOSBImage(ctx context.Context, mosb *mcfgv1.Mach
 	}
 
 	image := string(mosb.Spec.RenderedImagePushSpec)
-	isOpenShiftRegistry, err := b.isOpenShiftRegistry(image)
+	isOpenShiftRegistry, err := ctrlcommon.IsOpenShiftRegistry(context.TODO(), image, b.kubeclient, b.routeclient)
```
Comment (pablintino) [inline pkg/controller/build/reconciler.go]: I think we need to pass the caller's context here (from OSBuildController.ctx) instead of creating a new one.

The context is critical for propagating cancellation signals, deadlines, and loggers (among others). Using a new one here can lead to hard-to-debug issues, like requests not timing out or containers not terminating when the parent operation is cancelled.

File: pkg/server/cluster_server.go [hunk 10]
```diff
@@ -224,3 +264,116 @@ func kubeconfigFromSecret(secretDir, apiserverURL string, caData []byte) ([]byte
...
+		klog.Infof("Pool %s has MachineOSBuild but build not successful yet", pool.Name)
+		return ""
+	}
+
+	imageSpec := moscState.GetOSImage()
+
+	// Don't serve internal registry images during bootstrap because cluster DNS is not available yet
+	// New nodes will bootstrap with base image, then update to layered image after joining cluster (2 reboots)
+	// External registries (Quay, etc.) will still get the 1-reboot optimization
+	isInternal, err := ctrlcommon.IsOpenShiftRegistry(context.TODO(), imageSpec, cs.kubeclient, cs.routeclient)
```
Comment (pablintino) [inline pkg/server/cluster_server.go:364]: Related to my other comment. The MCS doesn't propagate a context and it seems there's no logic to handle a safe exit, so I guess using TODO here is fine.

File: pkg/controller/build/reconciler.go (hunk 9 above)
Comment (dkhater-redhat) [reply pkg/controller/build/reconciler.go]: good catch. thank you

File: pkg/server/cluster_server.go (hunk 5 above)
Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:350]: done!

**Comment by ptalgulk01** [comment]
### Verification Steps

#### Environment Setup
  - **OpenShift Version**: 4.21.0-0-2025-10-27-110611-test-ci-ln-8y8ddg2-latest
  - **Platform**: AWS
 
#### Pre-requisites

  #### 1. Create Image Registry Secret

  ```bash
  oc create secret docker-registry layering-push-secret \
  --docker-server=quay.io \
  --docker-username=<username> \
--docker-password=<password> \
  --docker-email="" \
  -n openshift-machine-config-operator
secret/layering-push-secret created
```
 2. Verify MCO Namespace
  ```bash
 $ oc get pods -n openshift-machine-config-operator
layering-push-secret                        kubernetes.io/dockerconfigjson        1      12s
```
Test Procedure

Step 1: Create MachineOSConfig
```bash
oc create -f - << EOF
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineOSConfig
metadata:
  name: worker
spec:
  machineConfigPool:
    name: worker
  imageBuilder:
    imageBuilderType: Job
  baseImagePullSecret:
    name: $(oc get secret -n openshift-config pull-secret -o json | jq "del(.metadata.namespace, .metadata.creationTimestamp, .metadata.resourceVersion, .metadata.uid, .metadata.name)" | jq '.metadata.name="pull-copy"' | oc -n openshift-machine-config-operator create -f - &> /dev/null; echo -n "pull-copy")
  renderedImagePushSecret:
    name:  layering-push-secret
  renderedImagePushSpec: "quay.io/mcoqe/layering:ocl"
      
EOF
machineosconfig.machineconfiguration.openshift.io/worker created
```
Step 2: Wait for MachineOSBuild
 
```bash
oc get machineosbuild
NAME                                      PREPARED   BUILDING   SUCCEEDED   INTERRUPTED   FAILED   AGE
worker-fc7c674c0a04ecb490deb569f7c852e7   False      False      True        False         False    15h
```

  Step 3:   Verify MCP update:
```bash
oc get mcp worker 
NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT   READYMACHINECOUNT   UPDATEDMACHINECOUNT   DEGRADEDMACHINECOUNT   AGE
master   rendered-master-456f017aacff4031b0f3ee2cd4f5f6f4   True      False      False      3              3                   3                     0                      18h
worker   rendered-worker-972335f65d1112207e042141f44da687   True      False      False      3              3                   3                     0                      18h

oc debug node/ip-10-0-48-250.us-east-2.compute.internal
sh-5.1# chroot /host
sh-5.1# cat /etc/machine-config-daemon/currentimage
quay.io/mcoqe/layering@sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644sh-5.1# cat /etc/macsh-5.1# cat /etc/machine-config-daemon/desiredImage
cat: /etc/machine-config-daemon/desiredImage: No such file or directory
sh-5.1# rpm-ostree status
State: idle
Deployments:
* ostree-unverified-registry:quay.io/mcoqe/layering@sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644
                   Digest: sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644
                  Version: 9.6.20251022-0 (2025-10-27T14:28:45Z)

```
  Step 4: Scale Up New Worker Node

```bash
$ oc get machinesets.machine.openshift.io -n openshift-machine-api 
NAME                                DESIRED   CURRENT   READY   AVAILABLE   AGE
ppt-2710b-22dwx-worker-us-east-2a   1         1         1       1           18h
ppt-2710b-22dwx-worker-us-east-2b   1         1         1       1           18h
ppt-2710b-22dwx-worker-us-east-2c   1         1         1       1           18h

$ oc scale --replicas 2  machinesets.machine.openshift.io -n openshift-machine-api ppt-2710b-22dwx-worker-us-east-2a
machineset.machine.openshift.io/ppt-2710b-22dwx-worker-us-east-2a scaled
```

Monitor the MCD logs
```bash
oc logs machine-config-daemon-c6js8 -f
...
I1028 06:00:42.733392    2628 daemon.go:1775] Current image: quay.io/mcoqe/layering@sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644
I1028 06:00:42.733400    2628 daemon.go:1776] Desired image: quay.io/mcoqe/layering@sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644
```

  Step 5: Verify First-Boot Behavior
  Access the new node during bootstrap (if possible via console/SSH):

  ```bash
oc debug node/ip-10-0-4-217.us-east-2.compute.internal -- chroot /host cat /etc/machine-config-daemon/currentimage 
Starting pod/ip-10-0-4-217us-east-2computeinternal-debug-xnqvs ...
To use host binaries, run `chroot /host`
quay.io/mcoqe/layering@sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644
Removing debug pod ...

oc debug node/ip-10-0-4-217.us-east-2.compute.internal -- chroot /host rpm-ostree status
Starting pod/ip-10-0-4-217us-east-2computeinternal-debug-mjslk ...
To use host binaries, run `chroot /host`
State: idle
Deployments:
* ostree-unverified-registry:quay.io/mcoqe/layering@sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644
                   Digest: sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644
                  Version: 9.6.20251022-0 (2025-10-27T14:28:45Z)

Removing debug pod ...
  ```
   Verify No Unnecessary Reboots

```bash
oc get events --field-selector involvedObject.name=ip-10-0-94-34.us-east-2.compute.internal --sort-by='.lastTimestamp'
LAST SEEN   TYPE     REASON                      OBJECT                                          MESSAGE
62m         Normal   Uncordon                    node/ip-10-0-94-34.us-east-2.compute.internal   Update completed for config rendered-worker-972335f65d1112207e042141f44da687 and node has been uncordoned
62m         Normal   NodeDone                    node/ip-10-0-94-34.us-east-2.compute.internal   Setting node ip-10-0-94-34.us-east-2.compute.internal, currentConfig rendered-worker-972335f65d1112207e042141f44da687 to Done
62m         Normal   ConfigDriftMonitorStarted   node/ip-10-0-94-34.us-east-2.compute.internal   Config Drift Monitor started, watching against rendered-worker-972335f65d1112207e042141f44da687
50m         Normal   ConfigDriftMonitorStopped   node/ip-10-0-94-34.us-east-2.compute.internal   Config Drift Monitor stopped
50m         Normal   AddSigtermProtection        node/ip-10-0-94-34.us-east-2.compute.internal   Adding SIGTERM protection
50m         Normal   Cordon                      node/ip-10-0-94-34.us-east-2.compute.internal   Cordoned node to apply update
50m         Normal   Drain                       node/ip-10-0-94-34.us-east-2.compute.internal   Draining node to update config.
47m         Normal   InClusterUpgrade            node/ip-10-0-94-34.us-east-2.compute.internal   Updating from oscontainer quay.io/mcoqe/layering@sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644
47m         Normal   OSUpdateStarted             node/ip-10-0-94-34.us-east-2.compute.internal   Updating to a target config with default kernel
47m         Normal   OSUpgradeSkipped            node/ip-10-0-94-34.us-east-2.compute.internal   OS upgrade skipped; new MachineConfig (rendered-worker-ce9ca102c6043c4e0b06d3c701fe9e9b) has same OS image (quay.io/mcoqe/layering@sha256:94a71773b23a961243d3c43a1c74d86ba0b9be0ddc8fc4482ae5f5b70246b644) as old MachineConfig (rendered-worker-972335f65d1112207e042141f44da687)
47m         Normal   RemoveSigtermProtection     node/ip-10-0-94-34.us-east-2.compute.internal   Removing SIGTERM protection
47m         Normal   Reboot                      node/ip-10-0-94-34.us-east-2.compute.internal   Node will reboot into config rendered-worker-ce9ca102c6043c4e0b06d3c701fe9e9b
46m         Normal   Uncordon                    node/ip-10-0-94-34.us-east-2.compute.internal   Update completed for config rendered-worker-ce9ca102c6043c4e0b06d3c701fe9e9b and node has been uncordoned
46m         Normal   NodeDone                    node/ip-10-0-94-34.us-east-2.compute.internal   Setting node ip-10-0-94-34.us-east-2.compute.internal, currentConfig rendered-worker-ce9ca102c6043c4e0b06d3c701fe9e9b to Done
46m         Normal   ConfigDriftMonitorStarted   node/ip-10-0-94-34.us-east-2.compute.internal   Config Drift Monitor started, watching against rendered-worker-ce9ca102c6043c4e0b06d3c701fe9e9b
```

  Step 6:  Delete the MOSC
```bash
$ oc delete machineosconfig worker
machineosconfig.machineconfiguration.openshift.io "worker" deleted
```
```bash
oc debug node/ip-10-0-2-167.us-east-2.compute.internal -- chroot /host rpm-ostree status
Starting pod/ip-10-0-2-167us-east-2computeinternal-debug-dvxwh ...
To use host binaries, run `chroot /host`
State: idle
Deployments:
* ostree-unverified-registry:registry.build10.ci.openshift.org/ci-ln-8y8ddg2/stable@sha256:bca4d514da5152f284c7ae974eeb33e8af845d599e9c699d6d8a0ea11a2f5819
                   Digest: sha256:bca4d514da5152f284c7ae974eeb33e8af845d599e9c699d6d8a0ea11a2f5819
                  Version: 9.6.20251022-0 (2025-10-22T15:07:01Z)

Removing debug pod ...

oc debug node/ip-10-0-2-167.us-east-2.compute.internal -- chroot /host cat /etc/machine-config-daemon/currentimage
Starting pod/ip-10-0-2-167us-east-2computeinternal-debug-zslkl ...
To use host binaries, run `chroot /host`

Removing debug pod ...
```

  Results Summary

  ✅ Success Criteria Met

  - New node pivoted to layered image during first boot
  - Only one reboot occurred
  - Node annotations show correct currentImage and desiredImage
  - Image digest matches MOSC/MOSB output
  - No additional reboots after initial provisioning

/label qe-approved

File: pkg/server/cluster_server.go [hunk 11]
```diff
@@ -163,7 +183,27 @@ func (cs *clusterServer) GetConfig(cr poolRequest) (*runtime.RawExtension, error
 
 	addDataAndMaybeAppendToIgnition(caBundleFilePath, cc.Spec.KubeAPIServerServingCAData, &ignConf)
 	addDataAndMaybeAppendToIgnition(cloudProviderCAPath, cc.Spec.CloudProviderCAData, &ignConf)
-	appenders := getAppenders(currConf, cr.version, cs.kubeconfigFunc, []string{}, "")
+
+	desiredImage := cs.resolveDesiredImageForPool(mp)
+	klog.Infof("desiredImage is found to be %s", desiredImage)
```
Comment (pablintino) [inline pkg/server/cluster_server.go:188]: nit: For a follow up. If we don't resolve to an image the function returns an empty string, that is not the most nice thing to add to a log. I'd prefer to conditionally log the message you already have if the string is not empty.

Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:188]: yes you're totally right i saw this in the testing i did yesterday and forgot to remove. i will remove in the follow up PR

File: pkg/server/cluster_server.go [hunk 12]
```diff
@@ -163,7 +183,27 @@ func (cs *clusterServer) GetConfig(cr poolRequest) (*runtime.RawExtension, error
...
+		},
+		appendInitialMachineConfig,
+		func(cfg *ign3types.Config, _ *mcfgv1.MachineConfig) error { return appendCerts(cfg, []string{}, "") },
+		// Inject desired image into the MC
+		appendDesiredOSImage(desiredImage),
+		// Must be last
+		func(cfg *ign3types.Config, mc *mcfgv1.MachineConfig) error {
+			return appendEncapsulated(cfg, mc, cr.version)
+		},
+	}
```
Comment (pablintino) [inline pkg/server/cluster_server.go:205]: nit: Uhmm, I don't like this approach too much as if at some point we need to touch the `getAppenders` function we need to touch this code here, and there's no warranty of that happening.
Why not an approach like this?
```go
  func getAppenders(currMachineConfig string, version *semver.Version, f kubeconfigFunc, certs []string, serverDir string, extra ...appenderFunc) []appenderFunc {
      appenders := []appenderFunc{
          // append machine annotations file.
          func(cfg *ign3types.Config, _ *mcfgv1.MachineConfig) error {
              return appendNodeAnnotations(cfg, currMachineConfig)
          },
          // append kubeconfig.
          func(cfg *ign3types.Config, _ *mcfgv1.MachineConfig) error { return appendKubeConfig(cfg, f) },
          // append the machineconfig content
          appendInitialMachineConfig,
          func(cfg *ign3types.Config, _ *mcfgv1.MachineConfig) error { return appendCerts(cfg, certs, serverDir) },
      }

      // Insert extra appenders in the middle
      appenders = append(appenders, extra...)

      // This has to come last!!!
      appenders = append(appenders, func(cfg *ign3types.Config, mc *mcfgv1.MachineConfig) error {
          return appendEncapsulated(cfg, mc, version)
      })

      return appenders
  }
```
With that in place, the already existing call in the bootstrap code doesn't need to change and here the difference is:
```go
appenders := getAppenders(currConf, cr.version, cs.kubeconfigFunc, []string{}, "", appendDesiredOSImage(desiredImage))

```

**Review by pablintino — APPROVED** [review]
I've added a few comments that are not blockers, specially given that QE already validated the current state of the PR.
/lgtm

File: pkg/server/cluster_server.go (hunk 12 above)
Comment (dkhater-redhat) [reply pkg/server/cluster_server.go:205]: totally agreed. addressing this in my next PR. tysm
